test/unit/Cardano/Wallet/BlockSyncerSpec.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE DataKinds #-}
+
+module Cardano.Wallet.BlockSyncerSpec
+    ( spec
+    ) where
+
+
+import Prelude
+
+import Cardano.Wallet.BlockSyncer
+    ( blockSyncerBracket )
+import Cardano.Wallet.Primitive
+    ( Block (..), BlockHeader (..), Hash (..) )
+import Control.Concurrent
+    ( forkIO, threadDelay )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Data.ByteString
+    ( ByteString, pack )
+import Data.IORef
+    ( IORef, newIORef, readIORef, writeIORef )
+import Test.Hspec
+    ( Spec, describe, it )
+import Test.Hspec.Expectations
+    ( shouldBe )
+import Test.QuickCheck
+    ( generate, vector, withMaxSuccess )
+import Test.QuickCheck.Gen
+    ( Gen, choose, vectorOf )
+
+
+import qualified Data.List as L
+import qualified Data.Set as Set
+
+spec :: Spec
+spec = do
+    describe "Block syncer downloads all blocks" $ do
+        it "check ticking function flow" $ withMaxSuccess 5 $ do
+
+            chunkSizesToTest <- generateBlockChunkSizes
+            tickTime <- generate $ choose (1,5)
+
+            putStrLn $
+                "        block chunks injected :"
+                ++ show chunkSizesToTest
+                ++ " ticking time: "
+                ++ show tickTime
+                ++ " seconds"
+
+            let evalTestTimeInSec = (L.length chunkSizesToTest + 1)*tickTime
+
+            consecutiveBlocks <- liftIO $ mkConsecutiveTestBlocks (sum chunkSizesToTest)
+            consumerData <- newIORef $ BlocksConsumed []
+            producerData <- newIORef $ BlocksToInject (chunkSizesToTest,consecutiveBlocks)
+
+            blockToIOaction <- writeToIORefAction consumerData consecutiveBlocks
+            let blockDelivery = pushNextBlocks producerData
+            _ <- forkIO $ blockSyncerBracket blockDelivery blockToIOaction tickTime
+            threadDelay (evalTestTimeInSec*1000*1000)
+
+            (BlocksConsumed consumedData) <- readIORef consumerData
+
+            consumedData `shouldBe` ((map fst . reverse) consecutiveBlocks)
```

Comment:
It is rather unclear to me what the tests are testing here :thinking: ? A comment or a longer scenario would definitely helps. ASCII diagrams are also very welcome.

Reply:
I wanted to check first *ExactlyOnce* delivery. When the ticking function gained not-consuming-the-same-block-guarrantee I added also **AtLeastOnce* delivery test